### PR TITLE
cylc pause

### LINKED ISSUE #3897: cli: play|pause|stop
Rename the existing workflow lifecycle commands to match the [rose suite-run migration proposal](https://github.com/cylc/cylc-admin/blob/master/docs/proposal-rose-suite-run.md#cli-changes) and implement the other CLI changes.

I.E. do what it says in the proposal under the following sections:

* `cylc play`
  * With the exception of the `--re-run` option - #3898
* `cylc pause`
* `cylc release`

### PR DESCRIPTION
**Pull request by MetRonnie**
These changes close #3897

- Implement `cylc pause`
- `cylc play` resumes a paused workflow
- Decouple `cylc hold/release` from workflow state, so they only hold/release tasks (or set/remove the hold cycle point for all tasks\*)

\*The `[scheduling]hold after cycle point`, `cylc hold --after` and `cylc play --hold-after` have remained pretty much the way they were, contrary to some of the discussion in #3897. However, they no longer affect the status of the task pool or scheduler, it just sets a point and if a spawned task is beyond that point it will be held
 
**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] (master branch) I have opened a documentation PR at https://github.com/cylc/cylc-doc/pull/204
- [x] No dependency changes.

### CODE REVIEW

**Comment by hjoliver** [comment]
(Does this branch include the `cylc play` changes?)

**Comment by MetRonnie** [comment]
Yes, will rebase onto master now that's been merged

**Comment by MetRonnie** [comment]
Before I go through any more tests, can I ask @oliver-sanders and/or @hjoliver to review the changes to `scheduler.py` so far? https://github.com/cylc/cylc-flow/pull/4076/commits/5a71c4d5f13edbc565dde00d566022a57f686ead#diff-90711b005fe27b9f894e6cd09f08816d847dcf853518894c6b7e2139784cbcb8

**Comment by oliver-sanders** [comment]
Looks sensible to me, nice.

The only thing I'm unsure about is whether is_paused should cause `should_process_tasks` to return `False` or whether it should block the task processing itself in `process_task_pool` itself as there are other things going on in that method.

**Comment by MetRonnie** [comment]
@hjoliver What do you say?
https://github.com/cylc/cylc-flow/blob/5a71c4d5f13edbc565dde00d566022a57f686ead/cylc/flow/scheduler.py#L1549-L1596

I'm not sure what the side effects of this method are

**Comment by hjoliver** [comment]
(I still need to think about that question, ran out of time today sorry)

**Comment by oliver-sanders** [comment]
Cross-posting from element, here's the `process_task_pool` method:

https://github.com/cylc/cylc-flow/blob/5a71c4d5f13edbc565dde00d566022a57f686ead/cylc/flow/scheduler.py#L1210-L1237

Here's my suggestion:

```
    def process_task_pool(self):
        """Process ALL TASKS whenever something has changed that might
        require renegotiation of dependencies, etc"""
        LOG.debug("BEGIN TASK PROCESSING")
        time0 = time()
        if self._get_events_conf(self.EVENT_INACTIVITY_TIMEOUT):
            self.set_suite_inactivity_timer()
        if not self.is_paused and self.stop_mode is None and self.auto_restart_time is None:
            # ^ skip all this stuff if paused
            # ...
        self.broadcast_mgr.expire_broadcast(self.pool.get_min_point())
        self.xtrigger_mgr.housekeep()
        self.suite_db_mgr.put_xtriggers(self.xtrigger_mgr.sat_xtrig)
        LOG.debug("END TASK PROCESSING (took %s seconds)" % (time() - time0))
```

This would allow the other things in that function to work as they currently would for a "held" workflow.

**Comment by hjoliver** [comment]
Yep, I agree with @oliver-sanders :+1:

**Comment by MetRonnie** [comment]
In this https://github.com/cylc/cylc-flow/blob/c6fd91b5bb2b02fbce849f4e7f0096d3db92c4da/cylc/flow/scheduler.py#L157-L165

`release_suite` and `release_tasks` are not used as command names, even on master, see https://github.com/cylc/cylc-flow/blob/a582f22c14baf8373d1e96b05833dd58abea5ddd/cylc/flow/network/resolvers.py#L650-L659

Should I change them to
```diff
diff --git a/cylc/flow/scheduler.py b/cylc/flow/scheduler.py
index 4bb79bfa0..027a2374c 100644
--- a/cylc/flow/scheduler.py
+++ b/cylc/flow/scheduler.py
@@ -156,8 +156,7 @@ class Scheduler:
     # Dependency negotiation etc. will run after these commands
     PROC_CMDS = (
-        'release_suite',
-        'release_tasks',
+        'release',
         'kill_tasks',
         'force_spawn_children',
```

**Comment by hjoliver** [comment]
> Should I change them to ...

I think so, yes, but want to look at the associated code tomorrow...

File: cylc/flow/scheduler.py
Comment (MetRonnie) [inline cylc/flow/scheduler.py:1236]: I'm not sure if the `if not self.is_paused` should also cover the block above @hjoliver 
```python
# Forget tasks that are no longer preparing for job submission.
self.pre_submit_tasks = [
    itask for itask in self.pre_submit_tasks if
    itask.waiting_on_job_prep
]
```

**Comment by hjoliver** [comment]
> > Should I change them to ...
> 
> I think so, yes, but want to look at the associated code tomorrow...

@MetRonnie - after looking at this I'm doing a bit of post-SoD refactoring that will remove the need for the PROC_CMDS tu...ple. Hopefully done today and you can rebase this on top of it.  Or, for the moment, just replace the `release_` strings as you suggested.

File: cylc/flow/scheduler.py
Comment (hjoliver) [reply cylc/flow/scheduler.py:1236]: Job prep completion happens asynchronously, via subprocesses, so it can complete even if the flow is paused.  On the other hand, nothing will submit anyway if paused ... so I don't think it matters much either way.  But, maybe cover that block too, just for a tiny efficiency advantage.

**Review by oliver-sanders — COMMENTED** [review]
Looks good, will be happy when we can pause a workflow without messing with the states of tasks within it.

Tested hold/release, play/pause, tui, no issues so far.

Good job spotting some of those left-over files in the test battery, thanks.

File: cylc/flow/scripts/release.py
Comment (MetRonnie) [reply cylc/flow/scripts/release.py:111]: > So does this mean that if you call the `release` mutation with an empty list of task globs it will release everything?

Yes

> I think `cylc release --all` is equivalent to `cylc release '*'` so could just stick with the glob expansion for this use case.

Not quite - `--all` will unset the hold point (if it was set), whereas `'*'` will not

File: cylc/flow/scheduler.py
Comment (MetRonnie) [reply cylc/flow/scheduler.py:1433]: What about things like this https://github.com/cylc/cylc-flow/blob/240b8f6613a0f7b726725444aef5f84b9f4e560d/tests/integration/test_examples.py#L139-L140

File: cylc/flow/scheduler.py
Comment (MetRonnie) [reply cylc/flow/scheduler.py:919]: Thing is, it can't accept any task globs when you use it to set the hold point by `self.command_hold(point=whatever)`

File: cylc/flow/network/resolvers.py
Comment (MetRonnie) [reply cylc/flow/network/resolvers.py:39]: Avoid unnecessary imports mainly, this will only import the Scheduler when mypy is run

Comment (oliver-sanders) [reply cylc/flow/network/resolvers.py:39]: I'm guessing we can't do:

```
self.schd: 'cylc.flow.scheduler.Scheduler' = schd
```

In order for a `Scheduler` object to have been passed to the `Resolvers` instance `cylc.flow.scheduler` must already be loaded in memory, even so it would be annoying if we had to write imports like this everywhere to keep mypy happy.

File: cylc/flow/scheduler.py
Comment (oliver-sanders) [reply cylc/flow/scheduler.py:1433]: I guess:

```diff
  # pump the scheduler's heart manually
- schd.release_runahead_tasks() 
+ schd.pool.release_runahead_tasks() 
```

File: cylc/flow/network/resolvers.py
Comment (MetRonnie) [reply cylc/flow/network/resolvers.py:39]: Imho writing out the full path like that would be more annoying than doing `if TYPE_CHECKING... import...`. Anyway, hopefully the usefulness of mypy outweighs its annoyances.

**Comment by MetRonnie** [comment]
Converting to draft until #4103 is merged, I guess?

File: cylc/flow/network/schema.py
Comment (oliver-sanders) [inline cylc/flow/network/schema.py:1527]: There is a special type for cycle points:

```suggestion
        point = CyclePoint(
```

Under the hood these types are all just strings but the types allow the UI to work out what "things" a mutation works with.

**Review by hjoliver — APPROVED** [review]
Approving after a somewhat rushed review, but LGTM :tada:

**Comment by oliver-sanders** [comment]
One unhappy test after the merge:

tests/f/reload/23-cycle-point-time-zone.t

**Review by oliver-sanders — APPROVED** [review]
All good on the end-to-end testing :+1:.

MyPy is a bit angry since recent changes?

**Comment by MetRonnie** [comment]
Yes, the attribute declarations like `pool: Optional[TaskPool] = None` are not playing well. I'm trying to figure out the best solution

File: cylc/flow/scheduler.py
Comment (MetRonnie) [inline cylc/flow/scheduler.py:177]: If making these attributes without defaults, they have to be before attributes with defaults in an `@dataclass` class

**Comment by MetRonnie** [comment]
Strange that `tests/f/platforms/07-localhost-set-if-not-in-globalcfg` is failing on MacOS consistently while passing on Linux.

```
ok 1 - 07-localhost-set-if-not-in-globalcfg-validate
not ok 2 - 07-localhost-set-if-not-in-globalcfg-run
ok 3 - job-grep-ok
```

Suite log:
```
2021-03-18T22:42:10 DEBUG - PT5M suite inactivity timer starts NOW: 2021-03-18T22:42:10
2021-03-18T22:42:10 DEBUG - END TASK PROCESSING (took 0.00095367431640625 seconds)
2021-03-18T22:47:11 WARNING - suite timed out after inactivity for PT5M
2021-03-18T22:47:11 ERROR - Suite shutting down - Abort on suite inactivity is set
2021-03-18T22:47:11 WARNING - Orphaned task jobs:
	* foo.1 (submitted)
```
There is no `job.err` for `foo`.

**Comment by oliver-sanders** [comment]
Sure.

**Comment by oliver-sanders** [comment]
That's because it sets `job runner = at` which we wouldn't expect to work on MacOS (the at daemon needs to be authenticated and enabled before it can work). Now why did this test not fail on master, or perhaps it did but no one noticed?

**Comment by MetRonnie** [comment]
Shouldn't be any need to re-run the tests

**Comment by oliver-sanders** [comment]
Fair'nough

**Review by oliver-sanders — APPROVED** [review]

**Pull request merged**
